### PR TITLE
initial import

### PR DESCRIPTION
**Pull request by jdimiduk**

### CODE REVIEW

File: AutoMad/autoMad.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+@author: dragon
+"""
+
+import random
+import re
+import wordGrabber
+import madLibFeeder
+import madDictionary
+import nltk
+import pandas as pd
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: It is good practice to create a [requirements.txt](https://pip.pypa.io/en/stable/reference/requirements-file-format/) file listing all the libraries you use that are not part of the standard library. This makes it much easier for someone else (or you on a different machine) to get started using the code.

File: AutoMad/autoMad.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Oct 22 13:14:59 2024
+
+@author: dragon
+"""
+
+import random
+import re
+import wordGrabber
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: I like following a practice of separating out library imports differentiating third party libraries and files from your package

``` python
import random
import re
# ...

import wordGrabber
import madLibFeeder
# ...
```
It can also be nice to organize imports alphabetically within each group just to make it easier to keep track of things.

[isort](https://pycqa.github.io/isort/) will do this for you automatically.

File: AutoMad/autoMad.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+ADJECTIVES = ["JJ","JJR","JJS","WDT"]
+ADVERBS = ["RB","RBR","RBS","RP","WRB"]
+SILLIES = ["UH"]
+UNKNOWN = ["FW","PDT","POS"]
+PREPOSITIONS = ["IN","TO","WP","WP$"]
+PRONOUNS = ["PRP","PRP$"]
+ARTICLES = ["DT"]
+NAMES = ["NNP","NNPS"]
+
+def posToCategory(pos):
```
Comment (tdimiduk) [inline AutoMad/autoMad.py:43]: Anywhere I see a long chain of `if`s I start looking for ways to do it with a dictionary instead. Here I think you could do something like

```
PARTS_OF_SPEECH = {
  'noun' = NOUNS,
  'verb' = VERBS'
  # ...
def posToGategory(pos):
  for cat, posList in PARTS_OF_SPEECH.items():
    if pos in posList:
       return cat
```

File: AutoMad/autoMad.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    if pos in PRONOUNS:
+        return 'pronoun'
+    if pos in ARTICLES:
+        return 'article'
+    if pos in NAMES:
+        return 'name'
+    return pos + "word Not Found"
+        
+def autoMad(story, percent, dictionary):
+    story.replace("'","'")
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: This looks like it won't do anything. It appears to be replacing `'` with `'` (a single quote with a single quote). If this is doing something tricky like replacing something that prints like a `'` but is actually a distinct character I'd encourage you to add a comment explaining that.

File: AutoMad/autoMad.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                    wordpart = wordTypes[1]    
+                    pos = wordpart[1]
+                    wordType = posToCategory(pos)
+                    newWord = dictionary.getWordInflect(wordType,pos)
+                    libout += ' ' + newWord
+                    print(wordpart[0] + ' ' + pos + ' ' + newWord)
+                    
+        else:
+            libout += word 
+    #madStory = madLibFeeder.wholeApple(libout, result)
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: nit: It's generally good practice not to check in commented out code. If this is something you use sometimes, it's better to and a flag argument to the function and put this behind and if

File: AutoMad/autoMad.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        return 'pronoun'
+    if pos in ARTICLES:
+        return 'article'
+    if pos in NAMES:
+        return 'name'
+    return pos + "word Not Found"
+        
+def autoMad(story, percent, dictionary):
+    story.replace("'","'")
+    words = re.split(r"(?<!')\b(?!')",story)
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: This regular expression is pretty hard to read (as most get if they are at all complex). I'd encourage a comment explaining what you are doing here in plain English.

File: AutoMad/ipython.html
Comment (tdimiduk) [inline AutoMad/ipython.html:1]: It looks like this file got checked in in spite of our attempts not to last night. You will probably want to delete it before merging this pr.

File: AutoMad/autoMad.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        return 'preposition'
+    if pos in PRONOUNS:
+        return 'pronoun'
+    if pos in ARTICLES:
+        return 'article'
+    if pos in NAMES:
+        return 'name'
+    return pos + "word Not Found"
+        
+def autoMad(story, percent, dictionary):
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: I'd encourage you to add type annotations to all of your functions. It will make it easier to keep track of what is being passed in (i.e. it took me a little while to figure out that the `dictionary` argument is probably an instance of your `madDictionary` class, and not a python `dict`)

Then I'd also encourage you to get in the habit of running [mypy](https://github.com/python/mypy) frequently, it will catch some errors for you (kind of like you would get by compiling a java program)

File: AutoMad/madDictionary.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+"""
+
+from dataclasses import dataclass
+import numpy as np
+import pandas as pd
+import random
+import wordFormer
+
+@dataclass
+class madDictionary:
```
Comment (tdimiduk) [inline AutoMad/madDictionary.py]: General practice most places I've seen is to name classes starting with a capital letter
```suggestion
class MadDictionary:
```

File: AutoMad/madDictionary.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    dictionary : pd.DataFrame
+    
+    def getWord(self, wordType):
+        for category in self.dictionary.columns:
+            if wordType == category:
+            #print(random.choice(df[category].dropna()))
+                return random.choice(list(self.dictionary[category].dropna()))
```
Comment (tdimiduk) [inline AutoMad/madDictionary.py:25]: You use `list(self.dictionary[category].dropna())` several times, it might be worth creating a helper method for that.

That would save a little duplication, allow you to give it a clear name, and decouple the rest of your code from the specific dataframe representation of the dictionary. I might also consider doing the same thing for `self.dictionary.columns`

File: AutoMad/madDictionary.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        return wordType
+        
+    def inCategory(self, word, category):
+        if self.dictionary[category].str.fullmatch(word.title()).any():
+            return True
+        return False
```
Comment (tdimiduk) [inline AutoMad/madDictionary.py]:
```
if ...: 
  return True
return False
```
Looks redundant to me

Can you replace this with
```suggestion
        return self.dictionary[category].str.fullmatch(word.title()).any()
```
or does the match return some kind of match object you actively don't want leaking elswhere?
Or can you do
```suggestion
        return bool(self.dictionary[category].str.fullmatch(word.title()).any())
```

File: AutoMad/wordFormer.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+"""def inflectCorrect(word, possum):
+    e = inflect.engine()
+    if possum in ['NNS','NNPS']:
+        return e.plural(word)
+    if possum == 'VBG':
+        return e.present_participle(word)
+    if possum == 'VBD':
+        return e."""
+
```
Comment (tdimiduk) [inline AutoMad/wordFormer.py]: This looks like old code
```suggestion
```

File: AutoMad/wordFormer.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        return e.plural(word)
+    if possum == 'VBG':
+        return e.present_participle(word)
+    if possum == 'VBD':
+        return e."""
+    
+def inflectCorrect(word, possum):
+    formed = pyinflect.getInflection(word,possum)
+    if formed is not None:
+        return formed [0]
```
Comment (tdimiduk) [inline AutoMad/wordFormer.py]:
```suggestion
        return formed[0]
```

File: AutoMad/wordFormer.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    else:
+        return word
+    
+print(inflectCorrect('boringly','RBR'))
```
Comment (tdimiduk) [inline AutoMad/wordFormer.py]:
```suggestion
```

It's generally bad practice run code like this at the top level (module import). This looks like the kind of thing you might want in a unit test though

File: AutoMad/wordFormer.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+"""def inflectCorrect(word, possum):
+    e = inflect.engine()
+    if possum in ['NNS','NNPS']:
+        return e.plural(word)
+    if possum == 'VBG':
+        return e.present_participle(word)
+    if possum == 'VBD':
+        return e."""
+    
+def inflectCorrect(word, possum):
```
Comment (tdimiduk) [inline AutoMad/wordFormer.py:9]: This function could use a docstring it's not obvious what it does without going and reading pyinflect's documentation (even just linking to said documentation would be helpful)

File: AutoMad/madDictionary.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    
+    def getWordInflect(self, wordType, possum):
+        for category in self.dictionary.columns:
+            if wordType == category:
+            #print(random.choice(df[category].dropna()))
+                newWord = random.choice(list(self.dictionary[category].dropna()))
+                newWord = wordFormer.inflectCorrect(newWord, possum)
+                return newWord
+        return wordType
```
Comment (tdimiduk) [inline AutoMad/madDictionary.py]: These variable names leave me a bit confused about the types of what is going on here. This function looks like it could return either a `Word` or a `WordType` which, to my haskell brain at least, look like they might be things of a different "kind". Type annotations and a docstring might help

File: AutoMad/autoMad.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    story.replace("'","'")
+    words = re.split(r"(?<!')\b(?!')",story)
+    libout = ""
+    for word in words:
+        if re.fullmatch("[A-Za-z'0-9]+", word) is not None and random.random()<.01*percent:            
+            if dictionary.inCategory(word, 'pokemon'):
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: You probably already know this, but special casing a `'pokemon'` here is a code smell to me (and is probably related to why your code hard crashes if that is missing.

File: AutoMad/autoMad.py (hunk 7 above)
Comment (tdimiduk) [inline AutoMad/autoMad.py]: I'm having a little trouble figuring out what this function does. From the name, I assumed it might be your top level function, but then it looks like at might be returning a word lib. Type annotations and/or a docstring might help me review it better

File: AutoMad/madFace.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+app = Flask(__name__)
+
+@app.route("/")
+def loadHome():
+    return render_template("home.html",stories=wordGrabber.getStories())
+
+@app.route("/story", methods = ['POST'])
+def story():
+    if request.form.get('stories') is not None:
+        pickedStory = request.form.get('stories')
```
Comment (tdimiduk) [inline AutoMad/madFace.py]:
```suggestion
    pickedStory = request.form.get('stories')
    if pickedStory is not None:
```
Or something to avoid doing the lookup in two different places

File: AutoMad/madFace.py [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        rawStory = wordGrabber.getStory(wordGrabber.getStoryByName(pickedStory))
+        if "|" in rawStory:
```
Comment (tdimiduk) [inline AutoMad/madFace.py]: This is kind of mysterious, I'd encourage a comment explaining what you are doing here

File: AutoMad/madLibFeeder.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+"""
+
+import wordPicker
+import madLibify
+import sillyWordMaker
+import re
+import wordGrabber
+#import nltk
+
+def getName(character, name):
```
Comment (tdimiduk) [inline AutoMad/madLibFeeder.py:15]: I think/hope type annotations would help me understand this function

File: AutoMad/wordGrabber.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+import numpy as np
+import pandas as pd
+import madDictionary
+
+# If modifying these scopes, delete the file token.json.
+SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly","https://www.googleapis.com/auth/documents"]
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:22]: You could consider including the `SCOPES` in the `token.json` you write out so that the code could check if they have changed and invalidate automatically. But given that you are not likely to change this that may be overkill.

File: AutoMad/wordGrabber.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+        with open("token.json", "w") as token:
+            token.write(creds.to_json())
+    return creds
+
+def getStoryByName(name):
+    result = getGoogleSheet(MADLIB_SHEET_ID,STORIES)    
+    values=result['values']
+    df = pd.DataFrame(values)
+    df.columns = df.iloc[0]
+    df = df[1:]
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:55]: Can you do something like

```suggestion
    df = pd.DataFrame(values[1:], columns=values[0])
```

File: AutoMad/wordGrabber.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+    for index,row in df.iterrows():
+        if row['storyname'] == name:
+            return row['urlString']
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:58]: This looks like a situation where it might not be worth converting to a dataframe at all. If the result you get back is something like a list of dicts, you could do

```
for  row in result['values']:
  if row['storyname'] == name:
     return row['urlString']
```

File: AutoMad/wordGrabber.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+    return None
+
+def getStories():
+    result = getGoogleSheet(MADLIB_SHEET_ID,STORIES)
+    values=result['values']
+    df = pd.DataFrame(values)
+    df.columns = df.iloc[0]
+    df = df[1:]
+    return df.storyname
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:67]: Can this be

```suggestion
    values=result['values']['storyname']
```

It looks like you may be falling into the pattern some data scientists do of reaching for pandas for everything. Pandas is great, and if it is doing some normalizing for you that you need for this, that is fine. If not, it makes the code harder to understand than it needs to be, and will waste a lot of compute cycles (not that that matters much)

File: AutoMad/autoMad.py (hunk 3 above)
Comment (tdimiduk) [reply AutoMad/autoMad.py:43]: or
```
PARTS_OF_SPEECH = {
  'noun': ["NN","NNS"]
  # ...
}

File: AutoMad/autoMad.py [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+CONJUNCTIONS =["CC"]
+NUMBER = ["CD","LS"]
+ADJECTIVES = ["JJ","JJR","JJS","WDT"]
+ADVERBS = ["RB","RBR","RBS","RP","WRB"]
+SILLIES = ["UH"]
+UNKNOWN = ["FW","PDT","POS"]
+PREPOSITIONS = ["IN","TO","WP","WP$"]
+PRONOUNS = ["PRP","PRP$"]
+ARTICLES = ["DT"]
+NAMES = ["NNP","NNPS"]
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]: I think you can delete a bunch of this now?

File: AutoMad/autoMad.py [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        if madDictionary.inCategory(word, category):
+            return madDictionary.getWord(category)
+    wordTypes=nltk.pos_tag(nltk.word_tokenize(word))
+    wordpart = wordTypes[0]    
+    pos = wordpart[1]
+    wordType = posToCategory(pos)
+    newWord = madDictionary.getWordInflect(wordType,pos)
+    wordOut = ''
+    wordOut += newWord
+    print(wordpart[0] + ' ' + pos + ' ' + newWord)
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]:
```suggestion
    wordOut = madDictionary.getWordInflect(wordType,pos)
    print(wordpart[0] + ' ' + pos + ' ' + wordOut) # Do you want this print? Should it be behind a `if debug` flag?
```

File: AutoMad/autoMad.py [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+    if len(wordTypes)>1: #just to handle contractions
+        wordpart = wordTypes[1]    
+        pos = wordpart[1]
+        wordType = posToCategory(pos)
+        newWord = madDictionary.getWordInflect(wordType,pos)
+        wordOut += ' ' + newWord
+        print(wordpart[0] + ' ' + pos + ' ' + newWord)
```
Comment (tdimiduk) [inline AutoMad/autoMad.py]:
```suggestion
   contractionWord = handleContraction(wordTypes)
   if contractionWord:
      return wordOut + " " + contractionWord
   return wordOut
 
 def handleContraction(wordTypes): 
    if len(wordTypes) > 1:
        wordpart = wordTypes[1]    
        pos = wordpart[1]
        wordType = posToCategory(pos)
        return madDictionary.getWordInflect(wordType,pos
```

File: AutoMad/madLibFeeder.py
Comment (tdimiduk) [inline AutoMad/madLibFeeder.py:1]: Not reviewing this file further since you said it is old and due a rewrite

File: AutoMad/madLibReader.py
Comment (tdimiduk) [inline AutoMad/madLibReader.py:1]: Should this file be deleted from this pr? It looks duplicative and old

File: AutoMad/madLibify.py
Comment (tdimiduk) [inline AutoMad/madLibify.py:1]: Is this an old file that got included by accident?

File: AutoMad/soupPourer.py [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+def getPage(url):
+    response = requests.get(url)
+    html_data = response.text
+    soup = BeautifulSoup(html_data,"html.parser")
+    return soup
+
+def madPage(url, sillyFactor):
+    soup=getPage(url)
+    dictionary = wordGrabber.getDictionary()
+    #text = soup.body.find_all(string=True)
```
Comment (tdimiduk) [inline AutoMad/soupPourer.py]:
```suggestion
```

File: AutoMad/storySaver.py [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+import numpy as np
+import pandas as pd
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+
+SCOPES = ["https://www.googleapis.com/auth/spreadsheets.readonly","https://www.googleapis.com/auth/documents"]
```
Comment (tdimiduk) [inline AutoMad/storySaver.py]: This scope looks the same as in the other google api file. If it is indeed the same, consider having only one copy somewhere and importing it both places

File: AutoMad/storySaver.py [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+SAVEDOC = "1Oha700Q9R_gGvVT2G2NYdL8Ixpr0xGwNVDJ3jMWOOhI"
+
+def saveStory(story):
+    story=story.replace("<br>","\n")+"\n\n\n"
```
Comment (tdimiduk) [inline AutoMad/storySaver.py:16]: Is this still appropriate with moving the `<br>` replacement into the web frontend?

File: AutoMad/storySaver.py [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = InstalledAppFlow.from_client_secrets_file(
+                "credentials.json", SCOPES
+            )
+            creds = flow.run_local_server(port=0)
+        # Save the credentials for the next run
+        with open("token.json", "w") as token:
+            token.write(creds.to_json())
```
Comment (tdimiduk) [inline AutoMad/storySaver.py]: I think you have a copy of this credential workflow in your other google file, consider pulling it out into a function they can both use

File: AutoMad/templates/story.html [hunk 31]
```diff
@@ -0,0 +1,28 @@
...
+</head>  
+<body>
+<p>
+<div class="story" id="story" width: 100px;>{{story}}</div>
+</p>
+<br>
+<!--<form method="POST" action="">-->
+<!--<input type="hidden" name="story" data-value="{{story}}">-->
+<button name="save" value ="Save" onclick = save()> Save </button>
+<!--</form>-->
```
Comment (tdimiduk) [inline AutoMad/templates/story.html]: nit: commented stuff. Do you mean to leave this in?

File: AutoMad/wordGrabber.py [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+        for tab in allTabs:
+           # Get the DocumentTab from the generic Tab.
+            document_tab = tab.get('documentTab')
+            doc_content = document_tab.get('body').get('content')
+            plainText += read_structural_elements(doc_content)
+        return plainText
+    
+        
+    except HttpError as err:
+        print(err)
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py]:
```suggestion
        return "".join(getTabText(tab) for tab in get_all_tabs(document)
    except HttpError as err:
        print(err)
        
def getTabText(tab):
      # Get the DocumentTab from the generic Tab.
      document_tab = tab.get('documentTab')
      doc_content = document_tab.get('body').get('content')
      return read_structural_elements(doc_content)
```

Let me know if I am pushing the functional/list comprehension style at you too hard. I find things much easier to follow if you are not mutating strings and separating out independent concerns into functions, but if that is not helpful at your current stage I can back off.

File: AutoMad/wordGrabber.py [hunk 33]
```diff
@@ -0,0 +1,147 @@
...
+        return None
+
+def getDictionary():
+    result = getGoogleSheet(MADLIB_SHEET_ID,WORDS)
+    values=result['values']
+    df = pd.DataFrame(values)
+    df.columns = df.iloc[0]
+    df = df[1:]
+    df = df.replace('', np.nan)
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:107]: does this need to use a dataframe?

File: AutoMad/wordGrabber.py [hunk 34]
```diff
@@ -0,0 +1,147 @@
...
+    dictionary = madDictionary.MadDictionary(df)
+    return dictionary
+    
+#code from google documentation to extract the text from a document
+def get_all_tabs(doc):
+  all_tabs = []
+  for tab in doc.get('tabs'):
+    all_tabs.append(tab)
+  return all_tabs
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py]:
```suggestion
def get_all_tabs(doc):
  return [tab for tab in doc.get('tabs')]
```

Or you don't really need this function at all, you could just inline the above where you need it.

File: AutoMad/wordGrabber.py [hunk 35]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def read_paragraph_element(element):
```
Comment (tdimiduk) [inline AutoMad/wordGrabber.py:111]: nit: most of your functions use `camelCase`, while it looks like the ones you copied in use `underscore_word_seperation`. Either is fine, but it could be worth standardizing on one. In most python codebases I've seen (that are not Java translations) underscores are more common.

File: AutoMad/wordPicker.py
Comment (tdimiduk) [inline AutoMad/wordPicker.py:1]: Is this an old file you didn't mean to include?

File: AutoMad/autoMad.py (hunk 24 above)
Comment (jdimiduk) [reply AutoMad/autoMad.py]: yes

File: AutoMad/wordPicker.py
Comment (jdimiduk) [reply AutoMad/wordPicker.py:1]: yes

File: AutoMad/storySaver.py (hunk 30 above)
Comment (jdimiduk) [reply AutoMad/storySaver.py]: yes, for the moment I am just calling the wordGrabber version of the creds function. This file should probably be just in wordgrabber and the file named something better (like google sheet connector) but this gets the duplication out for the moment.

File: AutoMad/wordGrabber.py (hunk 33 above)
Comment (jdimiduk) [reply AutoMad/wordGrabber.py:107]: no probably not, but for the moment it works. I will change this when I rework this area next, but it's not the worst code in here.

File: AutoMad/storySaver.py (hunk 29 above)
Comment (jdimiduk) [reply AutoMad/storySaver.py:16]: Yes in this case, I'm pulling from the front end so that I don't have to save story state. There is a chance someone could generate a story and then try to save it after a server reboot.

File: AutoMad/wordPicker.py
Comment (tdimiduk) [reply AutoMad/wordPicker.py:1]: You need to actually delete the file locally, commit the deletion and push. I'm still seeing this file in the repo

File: AutoMad/autoMad.py (hunk 26 above)
Comment (jdimiduk) [reply AutoMad/autoMad.py]: I'm going to leave this as is for now, the function is probably technically correct, but this is a case where things are likely to change and I don't think how I'm handling contractions is fully correct so I expect this to change more.
